src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs

```diff
@@ -8,42 +8,42 @@ namespace System.IO.MemoryMappedFiles
         {
-            fileSize = -1;
+            // if the length has already been fetched and it's more than 0 it's a regular file and there is no need for the FStat sys-call
+            isRegularFile = fileHandle is not null && (fileSize > 0 || IsRegularFile(fileHandle));
 
-            if (fileStream != null)
+            if (isRegularFile)
             {
-                Interop.Sys.FileStatus status;
-
-                int result = Interop.Sys.FStat(fileStream.SafeFileHandle, out status);
-                if (result != 0)
+                if (access == MemoryMappedFileAccess.Read && capacity > fileSize)
                 {
-                    Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
-                    throw Interop.GetExceptionForIoErrno(errorInfo);
+                    throw new ArgumentException(SR.Argument_ReadAccessWithLargeCapacity);
                 }
 
-                bool isRegularFile = (status.Mode & Interop.Sys.FileTypes.S_IFCHR) == 0;
+                // one can always create a small view if they do not want to map an entire file
+                if (fileSize > capacity)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(capacity), SR.ArgumentOutOfRange_CapacityGEFileSizeRequired);
+                }
 
-                if (isRegularFile)
+                if (access == MemoryMappedFileAccess.Write)
                 {
-                    fileSize = status.Size;
-                    if (access == MemoryMappedFileAccess.Read && capacity > status.Size)
-                    {
-                        throw new ArgumentException(SR.Argument_ReadAccessWithLargeCapacity);
-                    }
+                    throw new ArgumentException(SR.Argument_NewMMFWriteAccessNotAllowed, nameof(access));
+                }
+            }
 
-                    // one can always create a small view if they do not want to map an entire file
-                    if (fileStream.Length > capacity)
-                    {
-                        throw new ArgumentOutOfRangeException(nameof(capacity), SR.ArgumentOutOfRange_CapacityGEFileSizeRequired);
-                    }
+            static bool IsRegularFile(SafeFileHandle fileHandle)
+            {
+                Interop.Sys.FileStatus status;
 
-                    if (access == MemoryMappedFileAccess.Write)
-                    {
-                        throw new ArgumentException(SR.Argument_NewMMFWriteAccessNotAllowed, nameof(access));
-                    }
+                int result = Interop.Sys.FStat(fileHandle, out status);
+                if (result != 0)
+                {
+                    Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
+                    throw Interop.GetExceptionForIoErrno(errorInfo);
                 }
+
+                return (status.Mode & Interop.Sys.FileTypes.S_IFCHR) == 0;
```

Comment:
I realize this is pre-existing, and I probably wrote it, but... is this the right condition?

Reply:
It's not. It's checking whether it's not a character-oriented device file (so for a socket or pipe, it would return true).

Great catch!